File: Elixir/jewels_and_stones.exs
# https://leetcode.com/problems/jewels-and-stones/
# 771. Jewels and Stones

defmodule Solution do
  def num_jewels_in_stones(jewels, stones) do
    jewel_list = String.graphemes(jewels)
    stone_list = String.graphemes(stones)

    Enum.reduce(stone_list, 0,
      &case &1 in jewel_list do
        true -> &2 + 1
        _ -> &2
      end
      )
  end
end

jewels = "aA"
stones = "aAAbbbb"

Solution.num_jewels_in_stones(jewels, stones)
|> IO.puts()  # 3

File: Python/jewels_and_stones.py
# https://leetcode.com/problems/jewels-and-stones/
# 771. Jewels and Stones

def numJewelsInStones(jewels, stones):
    jewels_in_stones = 0
    for j in stones:
        if j in jewels:
            jewels_in_stones += 1
    return jewels_in_stones


# Or functional
def numJewelsInStones(jewels, stones):
    def compounder(x):
        if x in jewels:
            return 1
        return 0
    return sum(map(compounder, stones))

jewels = "z"
stones = "ZZ"


if __name__ == '__main__':
    print(numJewelsInStones(jewels, stones))